### PR TITLE
Make Averager serializable and constructible from options. 

### PR DESCRIPTION
**Pull request by osheamonn**
## Proposed changes

Adds functionality to the averager that I found I needed for the control systems. 

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/ControlSystem/Averager.cpp [hunk 1]
```diff
@@ -193,5 +193,35 @@ std::array<DataVector, DerivOrder + 1> Averager<DerivOrder>::get_derivs() const
...
+}
+
+template <size_t DerivOrder>
+bool operator==(const Averager<DerivOrder>& avg1,
+                const Averager<DerivOrder>& avg2) {
+  return (avg1.avg_tscale_frac_ == avg2.avg_tscale_frac_) and
+         (avg1.average_0th_deriv_of_q_ == avg2.average_0th_deriv_of_q_) and
+         (avg1.averaged_values_ == avg2.averaged_values_) and
+         (avg1.times_ == avg2.times_) and (avg1.weight_k_ == avg2.weight_k_) and
+         (avg1.tau_k_ == avg2.tau_k_);
```
Comment (knelli2) [inline src/ControlSystem/Averager.cpp]: Should you also check `raw_qs_`? I don't think there can be a case where all the timescales, weights, averaged values, etc. are the same and the `raw_qs_` are different, but I'm not 100% sure.

File: tests/Unit/ControlSystem/Test_Averager.cpp [hunk 2]
```diff
@@ -11,7 +11,10 @@
 
 #include "ControlSystem/Averager.hpp"
 #include "DataStructures/DataVector.hpp"
+#include "Framework/TestCreation.hpp"
+#include "Framework/TestHelpers.hpp"
 #include "Utilities/ConstantExpressions.hpp"
+#include "Framework/TestHelpers.hpp"
```
Comment (knelli2) [inline tests/Unit/ControlSystem/Test_Averager.cpp]: Included `TestHelpers` twice.

File: tests/Unit/ControlSystem/Test_Averager.cpp [hunk 3]
```diff
@@ -248,6 +251,28 @@ SPECTRE_TEST_CASE("Unit.ControlSystem.Averager.BadCallToAverageTime",
...
+SPECTRE_TEST_CASE("Unit.ControlSystem.Averager.EqualityAndSerialization",
+                  "[ControlSystem][Unit]") {
+  Averager<2> averager1(1.0, true);
+  Averager<2> averager2(1.0, true);
+  Averager<2> averager3(0.5, false);
+
+  CHECK(averager1 == averager2);
+  CHECK(averager1 != averager3);
+
+  CHECK(serialize_and_deserialize(averager1) == averager1);
```
Comment (knelli2) [inline tests/Unit/ControlSystem/Test_Averager.cpp]: You should also check that two averagers are (not)equal after an update. Here you're just checking that the default values of most members are the same.

Maybe also add a few `CHECK_FALSE`s as well.

File: src/ControlSystem/Averager.cpp [hunk 4]
```diff
@@ -193,5 +193,35 @@ std::array<DataVector, DerivOrder + 1> Averager<DerivOrder>::get_derivs() const
...
+bool operator!=(const Averager<DerivOrder>& avg1,
+                const Averager<DerivOrder>& avg2) {
+  return not(avg1 == avg2);
+}
+
 // explicit instantiations
 template class Averager<2>;
+template bool operator==(const Averager<2>& avg1, const Averager<2>& avg2);
+template bool operator!=(const Averager<2>& avg1, const Averager<2>& avg2);
+/// \endcond
```
Comment (knelli2) [inline src/ControlSystem/Averager.cpp]: Doxygen doesn't work in cpp files.

File: src/ControlSystem/Averager.hpp [hunk 5]
```diff
@@ -46,14 +67,23 @@ class Averager {
...
   ~Averager() = default;
 
+  template <size_t DDerivOrder>
+  friend bool operator==(const Averager<DDerivOrder>&,
+                         const Averager<DDerivOrder>&);
+
+  template<size_t DDerivOrder>
+  friend bool operator!=(const Averager<DDerivOrder>&,
+                         const Averager<DDerivOrder>&);
+
```
Comment (knelli2) [inline src/ControlSystem/Averager.hpp]: Clang-tidy complaining about duplicates. Maybe because they are public?

**Review by knelli2 — CHANGES_REQUESTED** [review]

File: src/ControlSystem/Averager.hpp [hunk 6]
```diff
@@ -8,6 +8,10 @@
 #include <deque>
 #include <optional>
 
+#include <pup.h>
+#include <pup_stl.h>
+
+#include "Parallel/PupStlCpp17.hpp"
 #include "DataStructures/DataVector.hpp"
```
Comment (wthrowe) [inline src/ControlSystem/Averager.hpp:13]: Alphabetize.

File: src/ControlSystem/Averager.hpp (hunk 5 above)
Comment (wthrowe) [reply src/ControlSystem/Averager.hpp]: It's a bug.  Just NOLINT them.

File: src/ControlSystem/Averager.hpp [hunk 7]
```diff
@@ -46,14 +50,23 @@ class Averager {
...
+  Averager& operator=(const Averager&) = default;
   ~Averager() = default;
 
+  template <size_t DDerivOrder>
+  friend bool operator==(const Averager<DDerivOrder>&,
+                         const Averager<DDerivOrder>&);
+
+  template<size_t DDerivOrder>
+  friend bool operator!=(const Averager<DDerivOrder>&,
+                         const Averager<DDerivOrder>&);
```
Comment (wthrowe) [inline src/ControlSystem/Averager.hpp]: This one doesn't need to be a friend.

File: src/ControlSystem/Averager.hpp [hunk 8]
```diff
@@ -36,6 +38,21 @@
...
+  };
+
+  struct Average0thDeriv {
+    using type = bool;
+    static constexpr Options::String help = {
+        "Whether to average the 0th derivative"};
+  };
+
+  using options = tmpl::list<AverageTimescaleFraction, Average0thDeriv>;
+  static constexpr Options::String help{"Averager"};
```
Comment (wthrowe) [inline src/ControlSystem/Averager.hpp]: Should probably put something here.

File: tests/Unit/ControlSystem/Test_Averager.cpp [hunk 9]
```diff
@@ -248,6 +251,28 @@ SPECTRE_TEST_CASE("Unit.ControlSystem.Averager.BadCallToAverageTime",
   averager.average_time(0.0);
 }
 
+SPECTRE_TEST_CASE("Unit.ControlSystem.Averager.EqualityAndSerialization",
+                  "[ControlSystem][Unit]") {
```
Comment (wthrowe) [inline tests/Unit/ControlSystem/Test_Averager.cpp]: Change the test cases in this file into functions and call them all from one test case.  We realized at some point that each test case has a significant overhead on some clusters.

File: src/ControlSystem/Averager.hpp [hunk 10]
```diff
@@ -8,7 +8,13 @@
 #include <deque>
 #include <optional>
 
+#include <pup.h>
+#include <pup_stl.h>
```
Comment (wthrowe) [inline src/ControlSystem/Averager.hpp:11]: Merge with above block of includes.

File: src/ControlSystem/Averager.cpp [hunk 11]
```diff
@@ -193,5 +193,35 @@ std::array<DataVector, DerivOrder + 1> Averager<DerivOrder>::get_derivs() const
...
+}
+
+template <size_t DerivOrder>
+bool operator!=(const Averager<DerivOrder>& avg1,
+                const Averager<DerivOrder>& avg2) {
+  return not(avg1 == avg2);
+}
+
 // explicit instantiations
 template class Averager<2>;
```
Comment (wthrowe) [inline src/ControlSystem/Averager.cpp:224]: This isn't new, but do you know why this is instantiated only for 2?

**Review by wthrowe — CHANGES_REQUESTED** [review]

**Review by wthrowe — DISMISSED** [review]
I'm happy for a squash, but wait for @knelli2's approval.

**Review by knelli2 — DISMISSED** [review]
LGTM :)

**Review by knelli2 — APPROVED** [review]

**Review by wthrowe — APPROVED** [review]

**Pull request merged**